### PR TITLE
Add mask optional arg to Batchnormalization.call()

### PR DESCRIPTION
**Pull request by ADH-LukeBollam**
Attempt to support masked values in BatchNormalization layer, so that mean and variance are correctly calculated ( #30161 )

Uses a modified version of nn.moments() which takes a `mask` argument.

@karmel

### CODE REVIEW

**Comment by gbaned** [comment]
@LukeBolly Please sign CLA in order to proceed with next steps. Thanks!

**Comment by LukeBolly** [comment]
Signed!

**Review by qlzh727 — CHANGES_REQUESTED** [review]

**Comment by LukeBolly** [comment]
changes have been implemented

File: tensorflow/python/keras/layers/normalization.py [hunk 1]
```diff
@@ -613,8 +613,50 @@ def _fake_update():
...
+      with ops.name_scope("moments", values=[inputs, axes]):
+          # The dynamic range of fp16 is too limited to support the collection of
+          # sufficient statistics. As a workaround we simply perform the operations
+          # on 32-bit floats before converting the mean and variance back to fp16
+          casted_inputs = math_ops.cast(inputs, dtypes.float32) if inputs.dtype == dtypes.float16 else inputs
+          # Compute true mean while keeping the dims for proper broadcasting.
+
+          if mask is not None:
+              mask = array_ops.expand_dims(mask, axis=-1)
+              binary_mask = math_ops.cast(mask, casted_inputs.dtype)
```
Comment (qlzh727) [inline tensorflow/python/keras/layers/normalization.py]: After the cast, it is not binary anymore, so probably rename it. Also, the mask is never used below, only the binary_mask is used. How about just:

mask = math_ops.cast(array_ops.expand_dims(mask, axis=-1), casted_inputs.dtype)

File: tensorflow/python/keras/layers/normalization_test.py [hunk 2]
```diff
@@ -216,6 +216,25 @@ def test_step(x, y):
...
+    normed = BatchNormalization(momentum=0.0)(masked)
+    model = tf.keras.models.Model(input, normed)
+    model.compile('rmsprop', 'mse',
+                  run_eagerly=testing_utils.should_run_eagerly())
+
+    model.fit(x=padded_data, y=padded_data, batch_size=10, epochs=5)
+    moments_mask = model.layers[2].get_weights()
+    model.fit(x=data, y=data, batch_size=10, epochs=5)
+    moments = model.layers[2].get_weights()
+    self.assertEqual(moments_mask[2][0], moments[2][0]) # moving average mean comparison
```
Comment (qlzh727) [inline tensorflow/python/keras/layers/normalization_test.py]: Assume the moments_mask[2] is the moving average mean here, why we only compare the first element which is [0] here?

File: tensorflow/python/keras/layers/normalization.py [hunk 3]
```diff
@@ -613,8 +613,50 @@ def _fake_update():
...
+                         inputs,
+                         axes,
+                         keep_dims=False,
+                         mask=None):
+
+      with ops.name_scope("moments", values=[inputs, axes]):
+          # The dynamic range of fp16 is too limited to support the collection of
+          # sufficient statistics. As a workaround we simply perform the operations
+          # on 32-bit floats before converting the mean and variance back to fp16
+          casted_inputs = math_ops.cast(inputs, dtypes.float32) if inputs.dtype == dtypes.float16 else inputs
```
Comment (qlzh727) [inline tensorflow/python/keras/layers/normalization.py]: i am curious about the fp16 issue here, since it is kind of orthogonal with regard to masking. I think we should separate this into another change.

**Review by qlzh727 — CHANGES_REQUESTED** [review]

File: tensorflow/python/keras/layers/normalization.py (hunk 3 above)
Comment (LukeBolly) [reply tensorflow/python/keras/layers/normalization.py]: This and a lot of the previous changes came directly from tf.nn.moments here:
https://github.com/tensorflow/tensorflow/blob/5e2a91f65cfb23f996136b9201d9312c9c36b941/tensorflow/python/ops/nn_impl.py#L1023
Would it make more sense to just modify this function to take a mask, instead of duplicating the code? It seems like we're creating a lot of extra stuff to maintain, when the functionality is almost identical.

Comment (qlzh727) [reply tensorflow/python/keras/layers/normalization.py]: Ah, I see. Thanks for for pointer. I think it actually make more sense to push the logic to tf.nn.moments with the new param of mask.

Comment (qlzh727) [reply tensorflow/python/keras/layers/normalization.py]: Also adding @alextp who is the general owner of the API since this will touch the tf.nn.moment API to have a new parameter.

Alex, masking is more like a Keras implementation detail for annotating the padded data. I am not very sure if we should push keras logic into lower level ops API.

Comment (alextp) [reply tensorflow/python/keras/layers/normalization.py]: I think an implicit mask in core TF APIs is a no-no. An explicit mask argument though seems useful in general.

**Comment by ialdencoots** [comment]
Will these updates also support masking when `renorm=True`?

File: tensorflow/python/keras/layers/normalization.py (hunk 3 above)
Comment (LukeBolly) [reply tensorflow/python/keras/layers/normalization.py]: where do we go from here @qlzh727 ?

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by ialdencoots** [comment]
Any updates as to the status of this?

**Comment by gbaned** [comment]
Hi Scott Zhu, Please help review this PR. Thanks!

**Comment by gbaned** [comment]
@LukeBolly,  @qlzh727, @alextp  Any update on this PR, please. Thanks!

File: tensorflow/python/keras/layers/normalization.py [hunk 4]
```diff
@@ -613,8 +613,50 @@ def _fake_update():
 
     return (r, d, out_mean, out_variance)
 
-  def _moments(self, inputs, reduction_axes, keep_dims):
-    mean, variance = nn.moments(inputs, reduction_axes, keep_dims=keep_dims)
+  def moment_calculation(self,
```
Comment (alextp) [inline tensorflow/python/keras/layers/normalization.py]: Can you make this a private method (_underscore in the name) so it doesn't become a part of TF's public API?

File: tensorflow/python/keras/layers/normalization.py (hunk 3 above)
Comment (alextp) [reply tensorflow/python/keras/layers/normalization.py]: Passing an explicit mask argument to nn.moments is reasonable to me.

**Review by alextp — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
@LukeBolly Could you please check alextp's comments and keep us posted. Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by LukeBolly** [comment]
I'll get to this shortly

**Comment by LukeBolly** [comment]
Fixed up the test and moved the mask handling into nn.moments, let me know if there is anything else needed.

**Comment by gbaned** [comment]
@LukeBolly Can you please resolve conflicts? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 18 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by qlzh727** [comment]
Closing this PR since it is quite stale. @LukeBolly feel free to reopen it if you are back and resolve the merge conflict.

**Comment by evancasey** [comment]
This would be really useful! Hoping this gets merged in 🙏
